src/librustc_typeck/check/mod.rs

```diff
@@ -1709,23 +1710,184 @@ fn check_opaque_for_cycles<'tcx>(
     span: Span,
     origin: &hir::OpaqueTyOrigin,
 ) {
-    if let Err(partially_expanded_type) = tcx.try_expand_impl_trait_type(def_id.to_def_id(), substs)
-    {
-        if let hir::OpaqueTyOrigin::AsyncFn = origin {
-            struct_span_err!(tcx.sess, span, E0733, "recursion in an `async fn` requires boxing",)
+    if let Err(partially_expanded_type) = tcx.try_expand_impl_trait_type(def_id.to_def_id(), substs) {
+        match (origin, &partially_expanded_type.kind) {
+            (hir::OpaqueTyOrigin::AsyncFn, _) => {
+                struct_span_err!(
+                    tcx.sess,
+                    span,
+                    E0733,
+                    "recursion in an `async fn` requires boxing"
+                )
                 .span_label(span, "recursive `async fn`")
                 .note("a recursive `async fn` must be rewritten to return a boxed `dyn Future`")
                 .emit();
-        } else {
-            let mut err =
-                struct_span_err!(tcx.sess, span, E0720, "opaque type expands to a recursive type",);
-            err.span_label(span, "expands to a recursive type");
-            if let ty::Opaque(..) = partially_expanded_type.kind {
-                err.note("type resolves to itself");
-            } else {
-                err.note(&format!("expanded type is `{}`", partially_expanded_type));
             }
-            err.emit();
+            (hir::OpaqueTyOrigin::FnReturn, ty::Opaque(..)) => {
+                let mut err = struct_span_err!(
+                    tcx.sess,
+                    span,
+                    E0720,
+                    "cannot resolve opaque return type `{}` to a concrete type",
+                    partially_expanded_type
+                );
+                err.span_label(span, "cannot resolve to a concrete type");
+                if let Some((hir_id, body_id)) = tcx
+                    .hir()
+                    .as_local_hir_id(def_id)
+                    .and_then(|hir_id| {
+                        let id = tcx.hir().get_parent_item(hir_id);
+                        tcx.hir().find(id).map(|n| (id, n))
+                    })
+                    .and_then(|(hir_id, node)| node.body_id().map(|b| (hir_id, b)))
+                {
+                    let tables = tcx.typeck_tables_of(tcx.hir().local_def_id(hir_id));
+                    let body = tcx.hir().body(body_id);
+                    let mut visitor = ReturnsVisitor::default();
+                    visitor.visit_body(body);
+                    if visitor
+                        .returns
+                        .iter()
+                        .filter_map(|expr| tables.node_type_opt(expr.hir_id))
+                        .map(|ty| {
+                            tcx.infer_ctxt().enter(|infcx| infcx.resolve_vars_if_possible(&ty))
```

Comment:
this doesn't really make sense -- it creates a *new inference context* (which will have a distinct set, initially empty, of inference variables) and then tries to resolve the inference variables present in `ty`. If there are any inference variables, then they come from a different inference context, so this would just cause ICEs and other errors if it does anything at all.

Reply:
however, since these tables come from `typeck_tables_of`, there will never be inference variables in its results anyway, so you can just remove the `map` call completely